pkg/build/strategies/sti/postexecutorstep.go

```diff
@@ -0,0 +1,464 @@
+				return fmt.Errorf("Couldn't rename %q -> %q: %v", old, new, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (step *downloadFilesFromBuilderImageStep) downloadAndExtractFile(artifactPath, artifactsDir, containerID string) error {
+	glog.V(3).Infof("Downloading file %q", artifactPath)
+
+	fd, err := ioutil.TempFile(artifactsDir, "s2i-runtime-artifact")
+	if err != nil {
+		return fmt.Errorf("Couldn't create temporary file for runtime artifact: %v", err)
+	}
+	defer func() {
+		fd.Close()
+		os.Remove(fd.Name())
+	}()
+
+	if err := step.docker.DownloadFromContainer(artifactPath, fd, containerID); err != nil {
+		return fmt.Errorf("Couldn't download file (%q -> %q) from container %s: %v", artifactPath, fd.Name(), containerID, err)
+	}
+
+	// after writing to the file descriptor we need to rewind pointer to the beginning of the file before next reading
+	if _, err := fd.Seek(0, os.SEEK_SET); err != nil {
+		return fmt.Errorf("Couldn't seek to the beginning of the file %q: %v", fd.Name(), err)
+	}
+
+	if err := step.tar.ExtractTarStream(artifactsDir, fd); err != nil {
+		return fmt.Errorf("Couldn't extract runtime artifact %q into the directory %q: %v", artifactPath, artifactsDir, err)
+	}
+
+	return nil
+}
+
+type startRuntimeImageAndUploadFilesStep struct {
+	builder *STI
+	docker  dockerpkg.Docker
+	fs      util.FileSystem
+}
+
+func (step *startRuntimeImageAndUploadFilesStep) execute(ctx *postExecutorStepContext) error {
+	glog.V(1).Info("Executing step: start runtime image and upload files")
+
+	fd, err := ioutil.TempFile("", "s2i-upload-done")
+	if err != nil {
+		return err
+	}
+	fd.Close()
+	lastFilePath := fd.Name()
+	defer func() {
+		os.Remove(lastFilePath)
+	}()
+
+	lastFileDstPath := "/tmp/" + filepath.Base(lastFilePath)
+
+	outReader, outWriter := io.Pipe()
+	defer outReader.Close()
+	defer outWriter.Close()
+
+	errReader, errWriter := io.Pipe()
+	defer errReader.Close()
+	defer errWriter.Close()
+
+	artifactsDir := filepath.Join(step.builder.config.WorkingDir, api.RuntimeArtifactsDir)
+
+	scripts := []string{api.Run}
+	useExternalAssembleScript := step.builder.externalScripts[api.AssembleRuntime]
+	if useExternalAssembleScript {
+		scripts = append(scripts, api.AssembleRuntime)
+	}
+
+	// We copy scripts to a directory with artifacts to upload files in one shot
+	for _, script := range scripts {
+		src := filepath.Join(step.builder.config.WorkingDir, api.UploadScripts, script)
+		dst := filepath.Join(artifactsDir, script)
+		glog.V(3).Infof("Copying file %q -> %q", src, dst)
+		if err := step.fs.Copy(src, dst); err != nil {
+			return fmt.Errorf("Couldn't copy file (%q -> %q): %v", src, dst, err)
+		}
+	}
+
+	image := step.builder.config.RuntimeImage
+	workDir, err := step.docker.GetImageWorkdir(image)
+	if err != nil {
+		return fmt.Errorf("Couldn't get working dir of %q image: %v", image, err)
+	}
+
+	commandBaseDir := workDir
+	if !useExternalAssembleScript {
+		// script already inside of the image
```

Comment:
can i provide assemble-runtime in my git repo in the .s2i/bin directory, like i can with assemble and run?